### ISSUE TITLE
Volumeattachments are not being removed on deleting pods

### ISSUE THREAD

**Issue by francisnijay1** [kubernetes/kubernetes#97230]
**What happened**: Volumeattachments are not being removed when a pod is deleted. This happens for some random storage classes based PVC while the remaining pods are having volumeattachments removed properly. 

**What you expected to happen**: Kubelet has to remove the volumeattachment which will make the external attacher(v3.0.0) to call ControllerUnpublish

**How to reproduce it (as minimally and precisely as possible)**: Install CSI Unity v1.3/v1.4 on k8s v1.19.x, there would be 3 storage classes would be generated during the driver installation. Create pvc and pods using all 3 storage classes. Now delete all pods one by one and one of the pods will have the volume attachment left out without being removed. The ATTACHED status of the volumeattachment stays true. This issue is intermittent and happens with one specific protocol based storageclass.

**Anything else we need to know?**:
After deleting the pod, CSI methods Node Unpublish volume and Node Unstage volume have been executed successfully and proper response is returned by the driver. I have extracted the below Controller managar logs from kube-system which says "DetachVolume.Detach succeeded for volume..." but still volumeattachment is left in the cluster.

I1211 04:54:08.682953       1 reconciler.go:203] attacherDetacher.DetachVolume started for volume "csi-vol-f-56f3de0a85" (UniqueName: "kubernetes.io/csi/csi-unity.dellemc.com^csi-vol-f-56f3de0a85-FC-apm0020*******-sv_4501") on node "domain**.lss.emc.com"
I1211 04:54:08.685970       1 operation_generator.go:1400] Verified volume is safe to detach for volume "csi-vol-f-56f3de0a85" (UniqueName: "kubernetes.io/csi/csi-unity.dellemc.com^csi-vol-f-56f3de0a85-FC-apm0020*******-sv_4501") on node "domain**.lss.emc.com"
I1211 04:54:08.688659       1 operation_generator.go:472] DetachVolume.Detach succeeded for volume "csi-vol-f-56f3de0a85" (UniqueName: "kubernetes.io/csi/csi-unity.dellemc.com^csi-vol-f-56f3de0a85-FC-apm0020*******-sv_4501") on node "domain**.lss.emc.com"

**Environment**:
- Kubernetes version (use `kubectl version`): v1.1.9.x
- Cloud provider or hardware configuration: CSI Unity 1.3/1.4
- OS (e.g: `cat /etc/os-release`):Centos 7.6, RHEL 7.6, 7.8
- Kernel (e.g. `uname -a`): Linux 3.10.0-957.27.2.el7.x86_64 #1 SMP Mon Jul 29 17:46:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug):
- Others:

/wg storage

**Comment by yuga711** [comment]
Did you get a chance to try on latest k8s version that has this issue fixed https://github.com/kubernetes/kubernetes/issues/77324?

**Comment by vermaa31** [comment]
Yes, tested in k8s - v1.20.2 and still, volume attachments are not getting deleted after pod deletion.

**Comment by yuga711** [comment]
On the KCM side, looking at waitForVolumeAttachDetachStatus, couldn't find a scenario where it may incorrectly report success during detach. When this issue happens, it would probably be worth to review the external-attacher logs.https://github.com/kubernetes/kubernetes/blob/dbf3c477a90e6b8136a3bbbb580ac6184c22998e/pkg/volume/csi/csi_attacher.go#L433

**Comment by vermaa31** [comment]
These are the attacher logs for a volume attachment that is not deleted even after its pod deletion. From these logs, it can be seen that the detach operation is never initiated for this volume attachment.
One more observation is: when a pod is deleted then the corresponding volume attachment is added with a deletion timestamp but here for this volume attachment, deletion timestamp is not added.
**Attacher Logs:**
`I0118 06:26:03.455194       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:03.455207       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:03.455217       1 csi_handler.go:243] Attaching "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:03.455228       1 csi_handler.go:422] Starting attach operation for "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:03.461106       1 csi_handler.go:304] VA finalizer added to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:03.461123       1 csi_handler.go:318] NodeID annotation added to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.464668       1 csi_handler.go:585] Saving attach error to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.470972       1 csi_handler.go:596] Saved attach error to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471009       1 csi_handler.go:226] Error processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68": failed to attach: rpc error: code = DeadlineExceeded desc = context deadline exceeded
I0118 06:26:18.471062       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471077       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471088       1 csi_handler.go:243] Attaching "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471100       1 csi_handler.go:422] Starting attach operation for "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471195       1 csi_handler.go:296] VA finalizer is already set on "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:18.471205       1 csi_handler.go:310] NodeID annotation is already set on "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.968379       1 csi_handler.go:585] Saving attach error to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972438       1 csi_handler.go:596] Saved attach error to "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972468       1 csi_handler.go:226] Error processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68": failed to attach: rpc error: code = Unknown desc =  runid=57 Export Volume Failed [{The requested operation has failed because the LUN has already been modified by another request. Please retry. (Error Code:0x6701500)}]
I0118 06:26:25.972510       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972527       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972537       1 csi_handler.go:243] Attaching "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972547       1 csi_handler.go:422] Starting attach operation for "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972548       1 controller.go:158] Ignoring VolumeAttachment "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" change
I0118 06:26:25.972633       1 csi_handler.go:296] VA finalizer is already set on "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:25.972651       1 csi_handler.go:310] NodeID annotation is already set on "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.113406       1 csi_handler.go:256] Attached "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.113436       1 util.go:37] Marking as attached "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.119859       1 util.go:51] Marked as attached "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.119884       1 csi_handler.go:262] Fully attached "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.119898       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.120034       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.120060       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.120073       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 06:26:27.120084       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.972720       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.972767       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:26:27.972785       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 06:26:27.972800       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:33:03.454682       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:33:03.454693       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:33:03.454709       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 06:33:03.454719       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:43:03.454825       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:43:03.454835       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:43:03.454919       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 06:43:03.454932       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:53:03.454943       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:53:03.454964       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 06:53:03.454978       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 06:53:03.455009       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:03:03.455293       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:03:03.455324       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:03:03.455347       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:03:03.455359       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:13:03.455579       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:13:03.455613       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:13:03.455626       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:13:03.455819       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:23:03.455808       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:23:03.455838       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:23:03.455947       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:23:03.455959       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:33:03.456013       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:33:03.456031       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:33:03.456044       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:33:03.456320       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:43:03.456328       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:43:03.456351       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:43:03.456454       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:43:03.456467       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:53:03.456600       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:53:03.456659       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 07:53:03.456740       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 07:53:03.457335       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:03:03.456900       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:03:03.457024       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:03:03.457052       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:03:03.457068       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:13:03.457311       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:13:03.457329       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:13:03.457342       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:13:03.457404       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:23:03.457141       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:23:03.457170       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:23:03.457183       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:23:03.457197       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:33:03.457462       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:33:03.457485       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:33:03.457502       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:33:03.457536       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:43:03.457781       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:43:03.457793       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:43:03.457805       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:43:03.457817       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:53:03.457935       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:53:03.457956       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 08:53:03.457972       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 08:53:03.458317       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:03:03.458216       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:03:03.458401       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:03:03.458415       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:03:03.458425       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:13:11.577269       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:13:11.577281       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:13:11.577291       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:13:11.577300       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:23:03.458573       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:23:03.458606       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:23:03.458621       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:23:03.458634       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:33:03.458842       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:33:03.458859       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:33:03.458988       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:33:03.458999       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:43:03.459153       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:43:03.459446       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:43:03.459465       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:43:03.459476       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:53:03.459853       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:53:03.459864       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 09:53:03.459953       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 09:53:03.459965       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:03:03.459701       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:03:03.459947       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:03:03.459960       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 10:03:03.460140       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:13:03.461084       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:13:03.461093       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:13:03.461104       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 10:13:03.461113       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:23:03.460289       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:23:03.460314       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:23:03.460337       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 10:23:03.460357       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:33:03.460444       1 controller.go:198] Started VA processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:33:03.460576       1 csi_handler.go:216] CSIHandler: processing VA "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"
I0118 10:33:03.460598       1 csi_handler.go:238] "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68" is already attached
I0118 10:33:03.460675       1 csi_handler.go:232] CSIHandler: finished processing "csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68"`

**kubectl describe volumeattachment csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68**
`Name:         csi-192ab83086c9f8a2188d331601dd8882d8be3c12c496b9350793d009d687ae68
Namespace:    
Labels:       <none>
Annotations:  csi.alpha.kubernetes.io/node-id: lg****9
API Version:  storage.k8s.io/v1
Kind:         VolumeAttachment
Metadata:
  Creation Timestamp:  2021-01-18T06:25:55Z
  Finalizers:
    external-attacher/csi-unity-dellemc-com
  Managed Fields:
    API Version:  storage.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:attacher:
        f:nodeName:
        f:source:
          f:persistentVolumeName:
    Manager:      kube-controller-manager
    Operation:    Update
    Time:         2021-01-18T06:25:55Z
    API Version:  storage.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:csi.alpha.kubernetes.io/node-id:
        f:finalizers:
          .:
          v:"external-attacher/csi-unity-dellemc-com":
      f:status:
        f:attached:
        f:attachmentMetadata:
          .:
          f:arrayId:
          f:host:
          f:lun:
          f:volumeContextId:
    Manager:         csi-attacher
    Operation:       Update
    Time:            2021-01-18T06:26:27Z
  Resource Version:  3633
  UID:               87871098-d73e-4b08-9004-8b6050d2dae8
Spec:
  Attacher:   csi-unity.dellemc.com
  Node Name:  lg*****9
  Source:
    Persistent Volume Name:  csivol-927867f283
Status:
  Attached:  true
  Attachment Metadata:
    Array Id:           ap**********1
    Host:               lg*******9
    Lun:                sv_6634
    Volume Context Id:  csivol-927867f283-iSCSI-ap***********1-sv_6634`

**Comment by dotkrnl** [comment]
/remove-lifecycle stale

still relative

**Comment by vermaa31** [comment]
We are still facing this issue. Logs are attached in the above comments.

**Comment by rockwyc992** [comment]
I found this bug in my cluster, too.  

And then, I change the `loglevel` of `kube-controller-manager` to the higher level, actually the `TraceAll` level in the OpenShift Cluster.
There is some wierd message as below:
``` 
I0611 01:34:14.135453       1 round_trippers.go:420] DELETE https://api-int.my-cluster.ocp:6443/apis/storage.k8s.io/v1/volumeattachments/csi-unity.dellemc.com%5Ecsiunity-**********-iSCSI-**************-sv_42
I0611 01:34:14.143541       1 round_trippers.go:446] Response Status: 404 Not Found in 8 milliseconds
I0611 01:34:14.143661       1 csi_attacher.go:411] kubernetes.io/csi: VolumeAttachment object [csi-unity.dellemc.com^csiunity-**********-iSCSI-**************-sv_42] for volume [csi-unity.dellemc.com^csiunity-**********-iSCSI-**************-sv_42] not found, object deleted
I0611 01:34:14.143673       1 operation_generator.go:472] DetachVolume.Detach succeeded for volume "csiunity-**********" (UniqueName: "csi-unity.dellemc.com^csiunity-**********-iSCSI-**************-sv_42") on node "logging03.dmz.starlux.ocp"
```  
 
With the function `Detach` in the [csi-attacher.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L400), when a detach request be sent to controller-manager, 
it would check the `volumeName` is a real `VolumeAttachment` or  a `UniqueName` in the CSI Plugin.  

But the check method is too simple to separate all of them, like below:
``` go
func isAttachmentName(unknownString string) bool {
    // 68 == "csi-" + len(sha256hash)  
    return strings.HasPrefix(unknownString, "csi-") && len(unknownString) == 68      
}
```

**Comment by rockwyc992** [comment]
In this case, we tried a workaround and succeeded to fix this problem perfectly.

## **TL;DR**

**Change your `--volume-name-prefix` more longer in the CRD `CSIUnity`** 

## Root Cause

First, there are something we need to know:
1. The `UniqueName` for a CSI volume is combined with `Driver Name` and `Volume ID (in CSI)`, or called `Volume Handle`.  
2. The `Volume Handle` is combined with `PV Name`, `Protocol`, `Array ID` and `Volume ID (in External Storage)`. 
3. The `PV Name` is generated automatically with the prefix what we defined in the sidecar `provisioner` of CRD `CSIUnity`.

Here is two mistake to hit this bug:
1. The `CSI Plugin` has a prefix `csi-` for the driver name, like `csi-unity.dellemc.com`.
  This is really a mistake with [dell/csi-unity](https://github.com/dell/csi-unity), but there is nothing we could do.
2.  The length of `UniqueName` is acctually **68 bytes**.  

In this case, changing the `Deiver Name` and `Protocol` is not possible, and changing `ArrayID` and `VolumeID` could affect other volume in the same external storage. 

## Resolusion

But we still be able to change the length of prefix of `PV Name` to promise total length is more than **68 bytes**. 
Increasing the length is better than decreasing, because the `Volume ID (in External Storage)` is incremental automatically.  

(Notice: You should backup all data in the origin volume, and provision new volume by re-creating the `PVC` to prevent this problem again)

Here is part of my CRD `CSIUnity`, see the `--volume-name-prefix` argument as below:
``` yaml
kind: CSIUnity
metadata:
  namespace: unity
  name: openshift-unity
spec:
  driver:
    common:
      image: dellemc/csi-unity:v1.4.0.000R
    configVersion: v3
    sideCars:
    - args:
      - --volume-name-prefix=csiunity-openshift
      image: k8s.gcr.io/sig-storage/csi-provisioner:v2.0.2
      imagePullPolicy: IfNotPresent
      name: provisioner
```

**Comment by rockwyc992** [comment]
@vermaa31 @francisnijay1 Please try this workaround, thanks.

**Comment by rockwyc992** [comment]
@oded-dd Have you try the workaround above?

**Comment by rockwyc992** [comment]
@oded-dd Please make sure the length of `Volume Name (for CSI)` is more than 68 byte.

Here is part of my CRD CSIUnity, see the `--volume-name-prefix` argument as below:

```
kind: CSIUnity
metadata:
  namespace: unity
  name: openshift-unity
spec:
  driver:
    common:
      image: dellemc/csi-unity:v1.4.0.000R
    configVersion: v3
    sideCars:
    - args:
      - --volume-name-prefix=csiunity-openshift
      image: k8s.gcr.io/sig-storage/csi-provisioner:v2.0.2
      imagePullPolicy: IfNotPresent
      name: provisioner
```

**Comment by oded-dd** [comment]
Sorry I didn't elaborate, we are experiencing the same issue with ebs controller. which does not have the same CRD
